Make Newlib usable with Synopsys MetaWare toolchain

abrodkin opened the pull request:
Synopsys MetaWare toolchain is a proprietary premium toolchain for ARC processors.
Now it's possible to build Newlib with MetaWare toolchain and use it as yet another standard C library for ARC coers.

To build it for ARC HS processor use the following commands, assuming MeratWare toolchain is in your `PATH` variable:
```
export AR_FOR_TARGET=llvm-arac
export AS_FOR_TARGET=asac
export CC_FOR_TARGET=ccac
export LD_FOR_TARGET=lldac
export RANLIB_FOR_TARGET=echo
export CFLAGS_FOR_TARGET="-av2hs -D__ARCHS__ -Dwint_t=int -DLONG_LONG_MAX=LLONG_MAX -DLONG_LONG_MIN=LLONG_MIN -DULONG_LONG_MAX=ULLONG_MAX -Hno_default_include -idirafter $METAWARE_ROOT/arc/inc"

mkdir build
cd build

../configure --target=arc --enable-multilib=no
make
```

shahab-vahedi commented:
@abrodkin please put the macro defs in this order:

```
#if defined (__GNUC__)
    ...
#elif defined (__CCAC__)
    ...
#else
    #error "Unknown compiler."
#endif
```

shahab-vahedi commented on libgloss/arc/crt0.S:
type in commit message: "chnage -> change"

shahab-vahedi commented on libgloss/arc/crt0.S:
could you do the same for `gcrt0.S` for the sake of consistency!?

abrodkin commented on libgloss/arc/arc-timer.c:
```diff
@@ -59,9 +61,9 @@ _arc_timer_read (unsigned int tim)
     {
       if (tim == 0)
-	return __builtin_arc_lr (ARC_TIM_COUNT0);
+	return read_aux_reg (ARC_TIM_COUNT0);
```
@shahab-vahedi @claziss what do you think? Fine by me but I usually don't like implicit redefinition which might mislead a reader, thus clearly introduced a `{read|write}_aux_reg()` thingy, which is equally non-native for both GCC & CCAC.

shahab-vahedi replied:
I totally get @evgeniy-paltsev 's point, but like you, I'd rather an explicit code that makes it clear what's happening. Masquerading MetaWare's builtins as GCC's is not something I favour.

evgeniy-paltsev replied:
Ok, fine by me :)

abrodkin commented:
> @abrodkin please put the macro defs in this order:
> 
> ```
> #if defined (__GNUC__)
>     ...
> #elif defined (__CCAC__)
>     ...
> #else
>     #error "Unexpected compiler"
> #endif
> ```

@shahab-vahedi that doesn't work that way because surprisingly `ccac` (i.e. `clang`) defines `__GNUC__`, see:
```
$ ccac -dM -E - < /dev/null | grep __GNUC__
#define __GNUC__ 4
```

So, will the following implementation be better than existing one?
```
#if defined (__CCAC__)
    ...
#elif defined (__GNUC__)
    ...
#else
    #error "Unexpected compiler"
#endif
```

shahab-vahedi commented:
> #if defined (__GNUC__)
>     ...
> #elif defined (__CCAC__)
>     ...
> #else
>     #error "Unexpected compiler"
> #endif

That is surprising indeed. I'll let @claziss make the final call on this. Till then, how about?

```
/* Our premium toolcahin has a bipolar disroder. */
#if defined (__GNUC__) && !defined (__CCAC__)
    ...
#elif defined (__CCAC__)
    ...
#else
    #error "Unexpected compiler"
#endif
```

abrodkin commented:
@shahab-vahedi pushed a re-spin with the following changes:
1. Fixed a commit msesage spello
2. Switched to `#if defined (xxx)` semantics in compiler checks
3. Removed `@` from `libgloss/arc/gcrt0.S` & `libgloss/arc64/crt0.S`

claziss commented:
```
#if defined (__CCAC__)
    ...
#elif defined (__GNUC__)
    ...
#else
    #error "Unexpected compiler"
#endif
```

@abrodkin Please comment why ``__CCAC__`` is used first (i.e.,llvm is defining both macros). @shahab-vahedi  I like your idea GNU first but the above is cleaner.

abrodkin commented:
@claziss done now. @BrunoASMauricio could you please run some testing on both ARC & ARC64 for this branch and if all is good, we may merge it, though not in `arc-2023.09` as we're done with that, but I guess `arc-2024.03` or something like that.

shahab-vahedi commented on libgloss/arc/arc-specific.h:
```diff
@@ -0,0 +1,30 @@
+#ifndef _ARC_SPECIFIC_H
+#define _ARC_SPECIFIC_H
+
+#ifdef __CCAC__
```
Please make the same changes as in `crt0.S` and `sbrk.c`.

shahab-vahedi commented:
Looks good to me with one [caveat](https://github.com/foss-for-synopsys-dwc-arc-processors/newlib/pull/52/files#r1404349532).

abrodkin commented:
> #error "Unexpected compiler"

Done now.

shahab-vahedi commented:
Thank you @abrodkin for all the reworks.

BrunoASMauricio commented:
@abrodkin newlib and GCC baremetal tests don't show regression with these changes
